### ISSUE TITLE
COMException with RunningTask.Refresh()

### ISSUE THREAD

**Issue by dahall** [dahall/TaskScheduler#212]
I want check task state and i write code
```
RunningTask rt = task.Run();
do
{
	Thread.Sleep(1000);
} while (rt.State == TaskState.Running);
```
However, I have infinite loop, because state is not updated (running always). When I change cycle body to 
```
do
{
	Thread.Sleep(1000);
	rt.Refresh(); // COMException here
} while (rt.State == TaskState.Running);
```
Code fires exception if state change from Running to Unknown.

It is correct behaviour?

P.S. In the end, i write  first cycle version with 
```
while (task.State == TaskState.Running);
```
and I do not have problems.

Originally posted: 2017-07-05T03:15:50

**Comment by dahall** [comment]
Please provide the details of the COMException. I suspect that in the second you are waiting, the task ends, changes it state, and then invalidates the COM object interface pointer, thus causing the exception.

Another option is to use TaskEventWatcher.

Originally posted: 2017-07-05T08:14:46

**Comment by dahall** [comment]
An unhandled exception of type 'System.Runtime.InteropServices.COMException' occurred in Microsoft.Win32.TaskScheduler.dll
HRESULT: 0x8004130B

Thanks for TaskEventWatcher, I will try it.

OT: How soon you move this project on github?

Originally posted: 2017-07-06T04:39:37

**Comment by dahall** [comment]
As I suspected that error code is SCHED_E_TASK_NOT_RUNNING: There is no running instance of the task. That confirms my suspicion I shared with you. I think you are much safer using TaskEventWatcher. There is an example in the Documentation area for using it.

I have already started moving to GitHub (https://github.com/dahall/TaskScheduler). The code is there as is most of the documentation. I haven't closed this site yet as I haven't found a solution for migrating the Discussions area.

Originally posted: 2017-07-06T06:43:16
